sync_user_permissions.py

```diff
@@ -0,0 +1,140 @@
+        self.ws_client = WorkplaceSearch(self.ws_host, http_auth=self.ws_token)
+        self.mapping_sheet_path = data.get("sharepoint_workplace_user_mapping")
+
+    def get_users_id(self):
+        """ This method returns the dictionary having the users as a key and it's unique id as a value
+        """
+        user_ids = {}
+        for collection in self.site_collections:
+            rel_url = f"{self.sharepoint_host}sites/{collection}/_api/web/siteusers"
+            response = self.sharepoint_client.get(rel_url, "?", "permission_users")
+            users = check_response(response.json(), "Could not fetch the SharePoint users.", "Error while parsing the response from url.", "sharepoint_users")
+            for user in users:
+                user_ids[user["Title"]] = user["Id"]
+        return user_ids
+
+    def get_user_groups(self, user_ids):
+        """ This method returns the groups of each user
+            :param user_ids: user ids to fetch the groups of the specific user
+        """
+        user_group = {}
+        for collection in self.site_collections:
+            rel_url = f"{self.sharepoint_host}sites/{collection}/"
+            for name, id in user_ids.items():
+                response = self.permissions.fetch_groups(rel_url, id)
+                groups = check_response(response.json(), "Could not fetch the SharePoint user groups.", "Error while parsing the response from url.", "user_groups")
+                user_group[name] = [group['Title'] for group in groups]
+        return user_group
+
+    def workplace_add_permission(self, permissions):
+        """ This method when invoked would index the permission provided in the paramater
+            for the user in paramter user_name
+            :param permissions: dictionary containing permissions of all the users
+        """
+        for user_name, permission_list in permissions.items():
+            try:
+                self.ws_client.add_user_permissions(
+                    content_source_id=self.ws_source,
+                    user=user_name,
+                    body={
+                        "permissions": permission_list
+                    },
+                )
+                logger.info(
+                    "Successfully indexed the permissions for user %s to the workplace" % (
+                        user_name
+                    )
+                )
+            except Exception as exception:
+                logger.exception(
+                    "Error while indexing the permissions for user: %s to the workplace. Error: %s" % (
+                        user_name, exception
+                    )
+                )
+
+    def sync_permissions(self):
+        """ This method when invoked, checks the permission of SharePoint users and update those user
+            permissions in the Workplace Search.
+        """
+        rows = {}
+        if (os.path.exists(self.mapping_sheet_path) and os.path.getsize(self.mapping_sheet_path) > 0):
```

Comment:
How is this `Sharepoint user` to `Workplace user` mapping managed?

Reply:
Hey, SharePoint user to Workplace user mapping is done using a csv file whose path is entered by the user in the config.yml file. The first entry in each row is the username in SharePoint and the second entry is the username in Workplace.

Reply:
So it's manually maintained?

Reply:
Yes, the user will manually fill in the mappings csv file which would have SharePoint to workplace user mapping. It is a part of configuration.

Reply:
Yes @wangch079 , so the user will have to fill in the mappings.csv file as a part of the configuration step which would have Sharepoint->Workplace user mappings before they start the connector. The connector will then automatically iterate over that CSV mappings and translate the users